### PR TITLE
Charge batteries in a way that maintains even charge level

### LINKED ISSUE #27019: Even Battery Charging.
**Is your feature request related to a problem? Please describe.**  
So, yes, you have TONS of batteries installed on your awesome mobile base. And you want to charge this one car battery/small storage battery you have on your cart/in your hand? Well, you can't..you gotta wait for the **TOP** battery to be done first! And how long will that take..? Oh I don't know, probably a couple of hours in-game..oh look, sun's down.

**Describe the solution you'd like**  
Kevin did mention this before, probably a year ago or so, that he would make batteries charge **evenly**, or rather, charge the lowest first. Shouldn't be too hard...right?

**Describe alternatives you've considered**  
Maybe the ability to rearrange battery slots? Or tank slots? That would be neat too.

**Additional context**  
Thanks again for making the Favorites tab! And again, just like the last suggestion I made, any other sorts of suggestions are more than welcome! Might even motivate devs to pick this suggestion up! c:

### PR DESCRIPTION
**Pull request by kevingranade**
#### Summary
```SUMMARY: None```

#### Purpose of change
Fixes #27019 
It's pretty arbitrary to charge whatever happens to be the first installed battery until it is full before charging any remaining batteries.  This change replaces that logic with charging logic that roughly maintains proportional charge level across all installed batteries.  Similar logic is applied to discharge logic to pull charges from the most charged battery.

#### Describe the solution
The code enumerates the batteries on the vehicle, and sorts them by charge level (percent of total charge) in a multimap.
Then it repeatedly pulls the least (or most) charged battery out of the map, charges (or discharges) enough charges to make the battery reach the next (or previous) charge level, and then re-inserts it into the map.  Repeatedly doing this will bring all batteries to the same charge level rapidly.

#### Describe alternatives you've considered
People have requested various options and switches for this feature, but this is a rough approximation of how charging occurs in a circuit, and it's a reasonable default.
I could see adding an action to connect or disconnect batteries from the charge/discharge circuit without uninstalling them, but I'm not interested in adding that feature right now.
There are a few additional tweaks that could be made for performance, but unless you have thousands of batteries and are charging/discharging huge quantities of power every turn, the single sweep over the parts list to find batteries should dominate execution time.

### CODE REVIEW

File: src/vehicle.cpp [hunk 1]
```diff
@@ -3762,14 +3762,28 @@ int vehicle::traverse_vehicle_graph( Vehicle *start_veh, int amount, Func action
...
-        }
-        if( p.is_available() && p.is_battery() ) {
-            int qty = std::min( long( amount ), p.ammo_capacity() - p.ammo_remaining() );
-            p.ammo_set( fuel_type_battery, p.ammo_remaining() + qty );
-            amount -= qty;
+    // Key parts by percentage charge level.
+    std::multimap<int, vehicle_part *> chargeable_parts;
+    for( vehicle_part &p : parts ) {
+        if( p.is_available() && p.is_battery() && p.ammo_capacity() > p.ammo_remaining() ) {
+            chargeable_parts.insert( { ( p.ammo_remaining() * 100 ) / p.ammo_capacity(), &p } );
```
Comment (BevapDin) [inline src/vehicle.cpp]: Just a minor detail, but one pair of parenthesis is not needed:
```C++
( a * b ) / c == a * b / c;
```

**Comment by kwl01skz** [comment]
When I use the original process to charge, If the gel generator is installed under the battery, it will automatically shut down due to lack of electricity, so it is difficult to generate electricity normally.

**Comment by kevingranade** [comment]
Is there a problem that hasn't been posted?  I don't know what a gel generator is, so you need more detail if you want me to look into that.

**Comment by ZhilkinSerg** [comment]
Merged to `0.D` (and `master` with with #28097).

**Pull request closed without merging**